File: rest/engine/engine/Engine.py
import random
from copy import deepcopy

ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
ncols = 5
nrows = 5

class Engine:
    def __init__(self, grid, letter = None) -> None:
        self.grid = grid
        self.letter = letter
        self.wordlist = self.get_wordlist("C:\\Users\\Filippo\\Desktop\\Beagle\\Parole Incrociate\\parole-incrociate\\rest\\engine\\engine\\word_storage\\280000_parole_italiane.txt")
    
    def show_grid_text(self, grid):
        """Output a text version of the filled grid wordsearch."""
        for irow in range(nrows):
            print(' | '.join(grid[irow]))

    def fill_grid_randomly(self, grid):
        """Fill up the empty, unmasked positions with random letters."""
        for irow in range(nrows):
            for icol in range(ncols):
                if grid[irow][icol] == ' ':
                    grid[irow][icol] = random.choice(ALPHABET)
                    return

    def is_grid_complete(self, grid):
            for row in grid:
                if ' ' in row:
                    return False
            return True

    def grid_as_svg(self, delta_grid, delta_letter):
        width, height = 1000, 1414
        """Output the SVG preamble, with styles"""

        svg =  """<?xml version="1.0" encoding="utf-8"?>
        <svg xmlns="http://www.w3.org/2000/svg"
            xmlns:xlink="http://www.w3.org/1999/xlink" width="{}" height="{}" >
        <defs>
        <style type="text/css"><![CDATA[
        line, path {{
        stroke: black;
        stroke-width: 4;
        stroke-linecap: square;
        }}fs
        path {{
        fill: none;
        }}

        text {{
        font: bold 24px Verdana, Helvetica, Arial, sans-serif;
        }}

        ]]>
        </style>
        </defs>
        """.format(width, height)

        """Return the wordsearch grid as a sequence of SVG <text> elements."""

        # A bit of padding at the top.
        YPAD = 20
        # There is some (not much) wiggle room to squeeze in wider grids by
        # reducing the letter spacing.
        letter_width = min(32, width / ncols)
        grid_width = letter_width * ncols
        # The grid is centred; this is the padding either side of it.
        XPAD = (width - grid_width) / 2
        letter_height = letter_width
        grid_height = letter_height * nrows
        s = []

        # Output the grid, one letter at a time, keeping track of the y-coord.
        y = YPAD + letter_height / 2
        for irow in range(nrows):
            x = XPAD + letter_width / 2
            for icol in range(ncols):
                letter = delta_grid[irow][icol]
                if letter != ' ':
                    s.append('<text x="{}" y="{}" text-anchor="middle">{}</text>'
                                    .format(x, y, letter))
                x += letter_width
            y += letter_height

        # We return the last y-coord uzed, to decide where to put the word list.
        svg = svg + '\n'.join(s) \
            + '<text x="{}" y="{}" text-anchor="middle" class="wordlist">{}</text>'.format(width * 0.25, 25, delta_letter) \
            + '</svg>'

        return svg

    def get_delta_matrix(self, original_grid, grid):
        delta_char = {
            "value": "",
            "location": (),
        }
        delta_grid = deepcopy(original_grid)
        for r in range(nrows):
            for c in range(ncols):
                if delta_grid[r][c] != grid[r][c] and delta_char['value'] == '':
                    delta_grid[r][c] = grid[r][c]
                    delta_char['value'] = grid[r][c]
                    delta_char['location'] = (r, c)
        return delta_grid, delta_char   


    def get_wordlist(self, wordlist_filename):
        """Read in the word list from wordlist_filename."""
        wordlist = []
        with open(wordlist_filename) as fi:
            for line in fi:
                # The word is upper-cased and comments and blank lines are ignored.
                line = line.strip().upper()
                if not line or line.startswith('#'):
                    continue
                wordlist.append(line)
        wordlist = [x for x in wordlist if len(x) <= 5 and len(x) > 1]
        random.shuffle(wordlist)
        wordlist = sorted(wordlist, key=lambda w: len(w), reverse=True)
        return wordlist

    def make_wordsearch(self, nrows, ncols, wordlist):
        """Attempt to make a word search with the given parameters."""

        def test_candidate(irow, icol, dx, dy, word):
            """Test the candidate location (icol, irow) for word in orientation
            dx, dy)."""                
            grid_word = ""
            mask_word = word
            for j in range(len(word)):
                if self.grid[irow][icol] not in (' ', word[j]):
                    return False
                grid_word = grid_word + self.grid[irow][icol]
                if self.grid[irow][icol] in word[j]:
                    mask_word = mask_word[:j] + '*' + mask_word[j + 1:]
                irow += dy
                icol += dx
                
            if word == grid_word:
                return False
            
            if self.letter not in mask_word:
                return False

            return True

        def place_word(word):
            # Left, down
            dxdy_choices = [(0,1), (1,0)]
            random.shuffle(dxdy_choices)
            for (dx, dy) in dxdy_choices:
                # Build a list of candidate locations for the word.
                candidates = []
                n = len(word)
                colmax = ncols - n if dx else ncols - 1
                rowmax = nrows - 1 if dx else nrows - n
                for irow in range(0, rowmax + 1):
                    for icol in range(0, colmax + 1):
                        if test_candidate(irow, icol, dx, dy, word):
                            candidates.append((irow, icol))
                # If we don't have any candidates, try the next orientation.
                if not candidates:
                    continue
                # Pick a random candidate location and place the word in this
                # orientation.
                irow, icol = random.choice(candidates)
                for j in range(n):
                    self.grid[irow][icol] = word[j]
                    irow += dy
                    icol += dx
                # We're done: no need to try any more orientations.
                break
            else:
                # If we're here, it's because we tried all orientations but
                # couldn't find anywhere to place the word. Oh dear.
                return False
            # print(word, loc, (dx, dy))
            return True

        # Iterate over the word list and try to place each word (without spaces).
        for word in wordlist:
            word = word.replace(' ', '').replace("'",'')
            if self.letter not in word:
                continue
            if place_word(word):
                return self.grid
    
        self.fill_grid_randomly(self.grid)
        return self.grid

    
    def calculate_results(self):

        results =  {
            "points": 0,
            "words": [],
        }

        def compute_points_by_direction(search_array, direction):
            for idx, word in enumerate(search_array):
                input = word
                stop_search = False
                while not stop_search:
                    count_mask_char = input.count('*')
                    filter_wordlist = [x for x in self.wordlist if len(x) <= len(x) - count_mask_char]
                    old_masked_chars = count_mask_char
                    new_masked_chars = count_mask_char
                    for word_to_search in filter_wordlist:
                        first_index = input.find(word_to_search)
                        if first_index >= 0:
                            results['words'].append({
                                "word": word_to_search,
                                "direction": direction,
                                "location": (idx, first_index) if direction == 'dx' else (first_index, idx)
                            })
                            len_word_to_search = len(word_to_search)
                            results['points'] = results['points'] + len_word_to_search
                            new_masked_chars = new_masked_chars + len_word_to_search
                            input = input.replace(word_to_search, '*'*len_word_to_search, 1)
                            
                    stop_search = True if new_masked_chars - old_masked_chars == 0 else False

        """Compute list of word in matrix"""
        words_rows = [''] * nrows
        words_columns = [''] * ncols
        for irow, row in enumerate(self.grid):
            for icol, col in enumerate(row):
                if not words_rows[irow]:
                    words_rows[irow] = ''
                words_rows[irow] = words_rows[irow] + col
                if not words_columns[icol]:
                    words_columns[icol] = ''
                words_columns[icol] = words_columns[icol] + col

        
        compute_points_by_direction(words_rows, "dx")
        compute_points_by_direction(words_columns, "dy")

        return results


    def compute_next_grid(self):
        original_grid = deepcopy(self.grid)
        new_grid = self.make_wordsearch(nrows, ncols, self.wordlist)
        delta_grid, delta_letter = self.get_delta_matrix(original_grid, new_grid)
        return delta_grid, delta_letter

File: rest/backend/src/engine/Engine.ts
import Utils from "../utils/Utils";


export type Matrix = Array<Array<string>>;

const ALPHABET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
const ncols = 5
const nrows = 5

export class Engine {
    grid: Array<Array<string>>;
    letter: {value: string, location: {dx: number, dy: number}};
    wordlist: string[];

    constructor(grid, letter = null) {
        this.grid = JSON.parse(JSON.stringify(grid));

        if (letter !== null) {
            this.letter = letter;
        }
        this.wordlist = this.getWordList("280000_parole_italiane.txt");
    }

    showGridText(grid: Matrix) {
        this.grid.forEach((row) => console.log(row.join(' | ')));
    }

    private fillGridRandomly(grid: Matrix, letter: string) {
        let firstCharPlaced = false
        for(let i = 0; i < grid.length; i++){
            for(let j = 0; j < grid[i].length; j++){
                if (grid[i][j] === " ") {
                    if(firstCharPlaced) {
                        grid[i][j] = ALPHABET.charAt(Math.floor(ALPHABET.length * Math.random()));
                        return;
                    }
                    else {
                        grid[i][j] = letter;
                        firstCharPlaced = true;
                    }
                }
            }
        }
    }

    private isGridComplete(grid: Matrix) {
        for(let i = 0; i < grid.length; i++){
            for(let j = 0; j < grid[i].length; j++){
                if (grid[i][j] === " ") {
                    return false
                }
            }
        }
        return true;
    }

    private getDeltaMatrix(originalGrid: Matrix, grid: Matrix, letter: string) {
        const deltaChar = {
            value: "",
            location: [],
        }
        const deltaGrid = JSON.parse(JSON.stringify(originalGrid));
        let letterAdded = false;
        for(let r = 0; r < grid.length; r++){
            for(let c = 0; c < grid[r].length; c++){  
                // add opponent letter selected
                if (deltaGrid[r][c] !== grid[r][c] && grid[r][c] === letter && !letterAdded) {
                    deltaGrid[r][c] = grid[r][c]
                    letterAdded = true
                }
                // add new letter selected
                if (deltaGrid[r][c] !== grid[r][c] && deltaChar['value'] === '') {
                    deltaGrid[r][c] = grid[r][c]
                    deltaChar['value'] = grid[r][c]
                    const loc = new Array<number>(2);
                    loc[0] = r;
                    loc[1] = c;
                    deltaChar['location'] = loc;
                }

            }
        }
        return [deltaGrid, deltaChar]
    }

    private getWordList(wordlistFilename: string) {
        const wordlist: string[] = []
        const fs = require('fs');
        const path = require('path');

        const p = __dirname + "/word_storage/" + wordlistFilename;

        const file = fs.readFileSync(p, {encoding:'utf8', flag:'r'})
        
        const lines = file.toString().split("\n")
        for (let line of lines) {
            //The word is upper-cased and comments and blank lines are ignored.
            let word: string = line.trim().toUpperCase();
            wordlist.push(word)
        }
        const wordlist_filter = wordlist.filter((val) => val.length <= 5 && val.length > 1);       
        
        const wordlist_shuffle = Utils.shuffle(wordlist_filter)
        return wordlist_shuffle.sort((a, b) => {
            return b.length - a.length;
        });
    
    };

    private testCandidate(irow: number, icol:number, dx: number, dy: number, word: string) {
        try {
        /* Tet the candidate location (icol, irow) for word in orientation
        dx, dy). */                
        let grid_word = "";
        let mask_word = word;
        for (let j=0; j < word.length; j++) {
            if (this.grid[irow][icol] !== " " && this.grid[irow][icol] !== word.charAt(j))
                return false
            grid_word = grid_word + this.grid[irow][icol]
            if (this.grid[irow][icol] === word.charAt(j))
                mask_word = mask_word.substring(0, j) + '*' + mask_word.substring(j + mask_word.length);
            irow += dy
            icol += dx
        }
        if (word === grid_word)
            return false;
        
        if (mask_word.indexOf(this.letter.value) < 0)
            return false;

        return true;
    }catch (e) {
        console.error(e);
    }
    }

    private placeWord(word: string): boolean {
        // Left, down
        const dxdy_choices = Utils.shuffle([{dx: 0, dy: 1}, {dx: 1, dy: 0}]);
        for (let {dx, dy} of dxdy_choices) {
            // Build a list of candidate locations for the word.
            let candidates = [];
            let n = word.length;
            let colmax = dx ? ncols - n : ncols - 1;
            let rowmax = dx ? nrows - 1 : nrows - n;
            for (let irow = 0; irow < rowmax + 1; irow++) {
                for (let icol = 0; icol < colmax + 1; icol++) {
                    if (this.testCandidate(irow, icol, dx, dy, word))
                        candidates.push({irow: irow, icol: icol})
                }
            }
            // If we don't have any candidates, try the next orientation.
            if (candidates.length === 0)
                continue
            // Pick a random candidate location and place the word in this
            // orientation.
            let {irow, icol} = candidates[Math.floor(candidates.length * Math.random())]
            for (let j = 0; j < n; j++) {
                this.grid[irow][icol] = word.charAt(j);
                irow += dy
                icol += dx
            }
            return true
        }
        
        // If we're here, it's because we tried all orientations but
        // couldn't find anywhere to place the word. Oh dear.
        return false
    }

    private makeWordSearch(nrows: number, ncols: number, wordlist: string[]) {
        
        for (let word of wordlist) {
            word = word.replace(' ','').replace("'",'');
            if (word.indexOf(this.letter.value) < 0)
                continue;
            if (this.placeWord(word))
                return this.grid;
        }

        this.fillGridRandomly(this.grid, this.letter.value);
        return this.grid;

    }

    private computePointsByDirection(results: any, search_array: string[], direction: string) {
        for (let wordIdx = 0; wordIdx < search_array.length; wordIdx++) {
            let input = search_array[wordIdx];
            let stop_search = false
            while (!stop_search) {
                let count_mask_char = (input.match(/\*/g) || []).length;
                let filter_wordlist = this.wordlist.filter((val) => val.length <= val.length - count_mask_char);
                let old_masked_chars = count_mask_char
                let new_masked_chars = count_mask_char
                for (let i=0; i < filter_wordlist.length; i++) {
                    let word_to_search = filter_wordlist[i];
                    let first_index = input.search(word_to_search)
                    if (first_index >= 0) {
                        results['words'].push({
                            "word": word_to_search,
                            "direction": direction,
                            "location": direction == 'dx' ? [wordIdx, first_index] : [first_index, wordIdx]
                        })
                        let len_word_to_search = word_to_search.length;
                        results['points'] = results['points'] + len_word_to_search
                        new_masked_chars = new_masked_chars + len_word_to_search
                        input = input.replace(word_to_search, '*'.repeat(len_word_to_search))
                    }
                }
                stop_search = new_masked_chars - old_masked_chars == 0 ? true : false;
            }
        }
    }

    calculateResults() {

        const results = {
            points: 0,
            words: []
        }

        // Compute list of word in matrix
        let words_rows = Array(nrows).fill('');
        let words_columns = Array(ncols).fill('');
        for (let irow = 0; irow < this.grid.length; irow++) {
            let row = this.grid[irow];
            for (let icol = 0; icol < row.length; icol++ ) {
                let col = row[icol];
                if (!words_rows[irow])
                    words_rows[irow] = ''
                words_rows[irow] = words_rows[irow] + col
                if (!words_columns[icol])
                    words_columns[icol] = ''
                words_columns[icol] = words_columns[icol] + col
            }
        }

        this.computePointsByDirection(results, words_rows, "dx")
        this.computePointsByDirection(results, words_columns, "dy")

        return results

    }

    computeNextGrid() {
        const originalGrid = JSON.parse(JSON.stringify(this.grid));
        const newGrid = this.makeWordSearch(nrows, ncols, this.wordlist);
        const [deltaGrid, deltaLetter] = this.getDeltaMatrix(originalGrid,newGrid, this.letter.value)
        const gridCompleted = this.isGridComplete(deltaGrid);
        return [deltaGrid, deltaLetter, gridCompleted]
    }

}